lgc/patch/PatchInOutImportExport.cpp

```diff
@@ -2783,120 +2840,13 @@ void PatchInOutImportExport::patchTcsBuiltInOutputExport(Value *output, unsigned
     break;
   }
   case BuiltInTessLevelOuter: {
-    if (builtInUsage.tessLevelOuter) {
-      // Extract tessellation factors
-      std::vector<Value *> tessFactors;
-      if (!elemIdx) {
-        assert(outputTy->isArrayTy());
-
-        unsigned tessFactorCount = 0;
-
-        auto primitiveMode = m_pipelineState->getShaderModes()->getTessellationMode().primitiveMode;
-        switch (primitiveMode) {
-        case PrimitiveMode::Isolines:
-          tessFactorCount = 2;
-          break;
-        case PrimitiveMode::Triangles:
-          tessFactorCount = 3;
-          break;
-        case PrimitiveMode::Quads:
-          tessFactorCount = 4;
-          break;
-        default:
-          llvm_unreachable("Should never be called!");
-          break;
-        }
-
-        for (unsigned i = 0; i < tessFactorCount; ++i) {
-          Value *elem = ExtractValueInst::Create(output, {i}, "", insertPos);
-          tessFactors.push_back(elem);
-        }
-
-        if (primitiveMode == PrimitiveMode::Isolines) {
-          assert(tessFactorCount == 2);
-          std::swap(tessFactors[0], tessFactors[1]);
-        }
-      } else {
-        assert(outputTy->isFloatTy());
-        tessFactors.push_back(output);
-      }
-
-      Value *tessFactorOffset = calcTessFactorOffset(true, elemIdx, insertPos);
-      storeTessFactorToBuffer(tessFactors, tessFactorOffset, insertPos);
-
-      assert(perPatchBuiltInOutLocMap.find(builtInId) != perPatchBuiltInOutLocMap.end());
-      unsigned loc = perPatchBuiltInOutLocMap[builtInId];
-
-      if (!elemIdx) {
-        // gl_TessLevelOuter[4] is treated as vec4
-        assert(outputTy->isArrayTy());
-
-        for (unsigned i = 0; i < outputTy->getArrayNumElements(); ++i) {
-          auto elem = ExtractValueInst::Create(output, {i}, "", insertPos);
-          auto elemIdx = ConstantInt::get(Type::getInt32Ty(*m_context), i);
-          auto ldsOffset = calcLdsOffsetForTcsOutput(elem->getType(), loc, nullptr, elemIdx, vertexIdx, insertPos);
-          writeValueToLds(elem, ldsOffset, insertPos);
-        }
-      } else {
-        auto ldsOffset = calcLdsOffsetForTcsOutput(outputTy, loc, nullptr, elemIdx, nullptr, insertPos);
-        writeValueToLds(output, ldsOffset, insertPos);
-      }
-    }
+    if (builtInUsage.tessLevelOuter)
```

Comment:
This isn't specific to your change, but why do we have this check here?

The function is responsible for writing TCS output builtins, and `builtInUsage` is the usage of TCS. So if this function is called for `builtInId == BuiltInTessLevelOuter`, then shouldn't `builtInUsage.tessLevelOuter` **always** be true here?

The same logic applies to BuiltInTessLevelInner.

If my reasoning here is correct, then you can remove the if-check and make the code slightly easier to follow by moving the lambda function back here into the switch, like:
```
case BuiltInTessLevelOuter:
case BuiltInTessLevelInner:{
  // Extract tessellation factors
  bool isOuter = builtInId == BuiltInTessLevelOuter;
  std::vector<Value *> tessFactors;
  if (!elemIdx) {
    assert(outputTy->isArrayTy());
    unsigned tessFactorCount = 0;
    const unsigned adjust = isOuter ? 2 : 0;
    ...
```
This allows us to read the code without jumping around so much.

Reply:
I see that you do a related change in #1089. I think removing the lambda function and doing everything in the switch statement would really be better.

Reply:
Yes, you're right. The check for `builtInUsage.tessLevelOuter` is unnecessary. Removed the checks and lambda function.